metadata-ingestion/src/datahub/api/circuit_breaker/assertion_circuit_breaker.py

```diff
@@ -0,0 +1,134 @@
+import logging
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Any, Dict, List, Optional
+
+from pydantic import Field
+
+from datahub.api.circuit_breaker.circuit_breaker import (
+    AbstractCircuitBreaker,
+    CircuitBreakerConfig,
+)
+from datahub.api.graphql import Assertion, Operation
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class AssertionCircuitBreakerConfig(CircuitBreakerConfig):
+    check_last_assertion_time: bool = Field(
+        default=True,
+        description="Whether to check if assertion happened after the dataset was last updated.",
+    )
+    time_delta: timedelta = Field(
+        default=(timedelta(days=1)),
+        description="In what timeframe should accept an assertion result if updated field is not available for the dataset",
+    )
+
+
+class AssertionCircuitBreaker(AbstractCircuitBreaker):
+    r"""
+    Datahub Assertion Circuit Breaker
+
+    The circuit breaker checks if there are passing assertion on the Dataset.
+    """
+    config: AssertionCircuitBreakerConfig
+
+    def __init__(self, config: AssertionCircuitBreakerConfig):
+        super().__init__(config.datahub_host, config.datahub_token, config.timeout)
+        self.config = config
+        self.assertion_api = Assertion(
+            datahub_host=config.datahub_host,
+            datahub_token=config.datahub_token,
+            timeout=config.timeout,
+        )
+
+    def get_last_updated(self, urn: str) -> Optional[datetime]:
+        operation_api: Operation = Operation(transport=self.assertion_api.transport)
+        operations = operation_api.query_operations(urn=urn)
+        if not operations:
+            return None
+        else:
+            return datetime.fromtimestamp(operations[0]["lastUpdatedTimestamp"] / 1000)
+
+    def _check_if_assertion_failed(
+        self, assertions: List[Dict[str, Any]], last_updated: Optional[datetime] = None
+    ) -> bool:
+        @dataclass
+        class AssertionResult:
+            time: int
+            state: str
+            run_event: Any
+
+        # If last_updated is set we expect to have at least one successfull assertion
+        if not assertions and last_updated:
+            return True
+
+        result: bool = True
+        assertion_last_states: Dict[str, AssertionResult] = {}
+        for assertion in assertions:
+            for run_event in assertion["runEvents"]["runEvents"]:
```

Comment:
We may want to ensure there is a runEvent for the assertion.

It's possible to have 0 run events

Reply:
(make sure to test this against URNs with no assertions defined at all)